Riot서버에서 최근 전적에 대하여 조회 및 DB Insert기능 구현

donsonioc2010 opened the pull request:
[#12 ]기능에 대한 구현입니다.
---

## 구현기능 정리
1. Puuid 파라미터를 수신 받아 최근 50경기의 MatchId를 List<String>으로 수신.
2. DB에서 MatchId를 조회하여 실제로 DB에 없는 MatchId만 List<String>으로 제작.
3. 존재하지 않는 MatchId를 기준으로 Riot에서 데이터를 수신해서 List<MatchDto>로 제작
4. List<MatchDto>를 순차적으로 DB에 데이터 Insert후 성공시 데이터 저장 Row에 대하여 Log로 남기는 작업입니다.

donsonioc2010 committed:
feat : #12개발을 위해 사용할 파일 추가

donsonioc2010 committed:
refactor : yml 잘못 commit하여 수정

donsonioc2010 committed:
feat : 게임정보를 조회하는 도메인, API추가 및 소환사조회 도메인 명칭변경, 그에따른 소스변, 테스트 추가

donsonioc2010 committed:
refactor : 오탈자 수정

donsonioc2010 committed:
Test : MatchId List 에 대한 테스트케이스 추가

donsonioc2010 committed:
refactor : HttpHeader Bean을 통한 관리 Edited

Riot서버와 통신을 진행할 때 현재 모두 GET통신을 진행하면 되는 상황이며,
요청하는 Header항목이 X-Riot-Token 한개뿐이어서 HttpHeader를 Bean으로 제작하여 공통으로
해당 Header를 사용하도록 코드 리팩토링

donsonioc2010 committed:
feat : Riot Server를 조회하여 최근 MatchId를 획득하는 기능 제작

donsonioc2010 committed:
feat, refactor : Dto추가 제작 및 DTO의 패키지 구조 수정.

1. Match와 관련된 Dto가 많다보니 관리를 쉽도록 구분하기 위하여 패키지 수정에 따른 수정코드 반영.
2. 게임 내용에 대하여 필요한 Dto제작.

donsonioc2010 committed:
refactor : Dto 수정

1. objectMapper -> restTemplate를 활용하여 바로 dto를 Return받도록 하기위한 어노테이션 추가
2. SQL문을 작성시에 필요한 속성 추가
3. 이름을 잘못 기입하여 못받아오던 속성 추가

donsonioc2010 committed:
feat : Riot서버에서 세부적인 게임정보 DB삽입(동기화)기능

1. MatchGameListService에서 최근 경기의 MatchId획득
2. 해당 리스트에서 DB에 존재하지 않는 MatchId만 분류
3. 분류한 MatchId를 다시 Riot에 요청하여 파싱한 MatchDto를 리스트에 추가
4. MatchDto가 모인 리스트를 DB에 순차적으로 Insert
5. 스키마의 경우 좀더 명시적인 표현이 필요한 컬럼 수정

donsonioc2010 committed:
refactor : Transactional 이슈파악을 위해 사용한 코드 Rollback 및 최근 경기의 MaxCount조정

- 최초 100건 테스트시 개발용 API-KEY Limit로 인한 조정

donsonioc2010 committed:
refactor : 추후 작업이 필요할 부분에 대하여 TODO추가

donsonioc2010 committed:
refactor : 로그 문구 수정과 변수 타입 수정

- Stream진행하면서 AtomicInteger 선언한 부분에 대하여 Integer로 수정
- Log문구 추가

donsonioc2010 committed:
test : Riot서버와 DB의 Sync에 대한 기능 완료

donsonioc2010 commented on src/main/java/com/nooblol/account/dto/match/MatchTeamDto.java:
```diff
@@ -0,0 +1,8 @@
+
+@NoArgsConstructor
+public class MatchTeamDto {
+
```
해당 Dto는 TeamDto로 사용을 하고 있어서 삭제가 필요한데 Commit을 누락한것 같습니다.

donsonioc2010 commented on src/main/java/com/nooblol/account/service/impl/MatchGameInfoServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * Puuid를 기반으로 Riot서버에서 최근 100게임의 MatchId List를 받아와서
```
최근 100게임으로 하려고 하였으나, 실제적으로 Local에서도 직접 테스트해본 결과 Limit에 걸려 최근 50경기로 수정하였습니다.

해당 주석은 수정하도록 하겠습니다.

donsonioc2010 commented on src/main/java/com/nooblol/account/service/impl/MatchGameInfoServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+
+    return matchIdList.stream()
+        .filter(matchId -> !existsMatchIdList.contains(matchId)).collect(Collectors.toList());
+  }
```
무조건적으로 DB를 Select하는 경우밖에 없어 readOnly를 사용하였습니다.

실제 조회를 하게되는 테이블의 경우 데이터가 가장 많을 테이블이어서 조금이라도 속도를 높여보고자 해당 성능을 사용했습니다.

donsonioc2010 commented on src/main/java/com/nooblol/account/service/impl/MatchGameInfoServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+
+  @Override
+  @Transactional(propagation = Propagation.REQUIRES_NEW)
+  public boolean insertMatchDataByDB(MatchDto matchDto) throws Exception {
+    MatchGameInfoDto infoDto = matchDto.getInfo();
+    if (infoDto == null) {
+      log.error("[InsertMatchData Fail] : MatchDto Is Null");
+      return false;
+    }
+    matchGameInfoMapper.insertMatchGameInfo(infoDto);
+    matchGameInfoMapper.insertMatchGameBans(infoDto.getTeams());
+    matchGameInfoMapper.insertMatchGameParticipants(infoDto);
+    matchGameInfoMapper.insertMatchGameUseStatRunes(infoDto);
+
+    Map infoMap = new HashMap();
+    infoMap.put("matchId", infoDto.getMatchId());
+    infoMap.put("participants", infoDto.getParticipants());
+    matchGameInfoMapper.insertMatchGameUseStyleRunes(infoMap);
+    return true;
+  }
```
해당 한건을 한개의 트랜잭션으로 묶어서 일괄처리, 일괄롤백을 하고 싶은데 이걸 처리하는 방법을 도저히 못찾아서 리뷰를 남깁니다.

원하는 기능은 다음과 같습니다.
- 50건의 List<MatchDto>를 받으면 `50건중에서 1건만 실패`할 경우 `해당 1건만`롤백이 되도록 트랜잭션 처리 하도록 희망
---
현재 
- 50건중 한건만이 아닌 50건이 전체 롤백되는 중

f-lab-kyle replied:
트랜잭션 범위를 나누면 되지 않을까요? 그리고 트랜젝션이 언제 발생(언제 롤백해야 하는지)를 명시적으로 어노테이션에 적어주는 것도 좋을것 같아요.

donsonioc2010 commented on src/main/resources/constants.yml:
```diff
@@ -1,8 +1,11 @@
+
+  match-domain: https://asia.api.riotgames.com/
+  match-list-search-by-puuid: lol/match/v5/matches/by-puuid/{puuid}/ids
+  match-game-info-by-match-id: lol/match/v5/matches/{matchId}
```
전적을 검색하는 도메인과 , 소환사 계정정보를 조회하는 도메인이 서로 달라서 구문을 위하여 수정

donsonioc2010 commented on src/main/java/com/nooblol/global/config/RiotConfiguration.java:
```diff
@@ -16,11 +18,25 @@ public class RiotConfiguration {
+
+  private String matchGameInfoByMatchId;
+
+  @Bean
+  public HttpHeaders initRiotHeader() {
+    HttpHeaders httpHeaders = new HttpHeaders();
+    httpHeaders.add("X-Riot-Token", getApiKey());
+
+    return httpHeaders;
```
해더에 들어가는 항목이 모두 `X-Riot-Token` 한개뿐이라 `HttpHeaders`를 `Bean`으로 제작하고 해당 Header를 가져오도록 수정하였습니다.

f-lab-kyle commented on NoobLoL.iml:
```diff
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
iml 파일이 repo에 추가되야 할 이유가 있을까요~?

f-lab-kyle commented on src/main/java/com/nooblol/account/controller/MatchGameController.java:
```diff
@@ -0,0 +1,21 @@
+@RestController
+@RequestMapping("/match")
+@RequiredArgsConstructor
+public class MatchGameController {
```
임의로 껍데기(?)만 만들어 두신거죠? 이런 경우는 가급적 PR에는 포함 안시키는게 좋을것 같아요~
커밋/푸시를 할때 어떤것들이 포함되는지 점검하는 습관도 중요할것 같습니다 :D

donsonioc2010 replied:
넵 알겠습니다 :D

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/MatchGameInfoServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+  public ResponseDto syncRiotToDbDataProcess(String puuid) throws Exception {
+    ResponseDto getMatchListData = matchGameListService.getMatchListId(puuid);
+
+    if (getMatchListData.getResultCode() == HttpStatus.OK.value()) {
```
이 조건을 반대로 하면 아래 인덴트를 줄일수 있지 않을까요??
e.g.

```java
if (getMatchListData.getResultCode() != HttpStatus.OK.value()) {
  return getMatchListData;
}
List<String> riotMatchIdList = (ArrayList<String>) getMatchListData.getResult();
...
```

donsonioc2010 replied:
인덴트를 줄여본다는 생각은 못해봤는데 생각을 해보니 가독성이 확실히 달라지는 것 같습니다.

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/MatchGameInfoServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+      List<MatchDto> inputMatchList = new ArrayList<>();
+
+      //MatchId를 기반으로 Riot에서 서버에서 상세 데이터를 받아 List에 추가
+      notExistsMatchList.stream().forEach(matchId -> {
```
notExistsMatchList 이  null인 경우는 없을까요? (npe)

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/MatchGameInfoServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+    if (response != null) {
+      return (MatchDto) response.getBody();
+    }
+    return null;
```
리턴을  null 하는 경우는 옵셔널을 활용해보세요. (사용처에서 npe방어를 명시적으로~)

f-lab-kyle commented on src/main/java/com/nooblol/account/service/impl/SummonerHistoryServiceImpl.java:
```diff
@@ -87,11 +89,9 @@ public ResponseDto selSummonerHistoryByRiot(String summonerId) {
-
-    return restTemplate.exchange(url, HttpMethod.GET, new HttpEntity<String>(httpHeaders),
-        String.class);
+    return restTemplate.exchange(
```
restTemplate 를 쓴 이유가 무엇일까요? 잘못되었다는건 아니고요 ㅎㅎ
webclient 도 있는데  restTemplate는 점점 없어질거라서요
https://github.com/spring-guides/gs-consuming-rest/issues/28

f-lab-kyle commented on src/main/resources/mybatis/mapper/account/MatchGameInfoMapper.xml:
```diff
@@ -0,0 +1,88 @@
+    <choose>
+      <when test="list.size != 0">
+        select match_id from match_gameinfo main
+        where exists(
+        select 1 from match_gameinfo sub
+        where main.match_id = sub.match_id
+        and match_id in
+        <foreach collection="list" item="item" index="index" separator="," open="(" close=")">
+          #{item}
+        </foreach>
+        )
```
연산은 밖에서 하는게 좋을것 같아요. DB에는 최소한의 부하만~

f-lab-kyle commented on src/main/resources/mybatis/mapper/account/MatchGameInfoMapper.xml:
```diff
@@ -0,0 +1,88 @@
+    INSERT INTO MATCH_GAME_RUNES(puuid, match_id, type, sort_no, perk)
+    VALUES
+    <foreach collection="participants" item="participants" separator=",">
+      (#{participants.puuid}, #{matchId}, 'stat_offense', 1,
+      #{participants.perks.statPerks.offense}),
+      (#{participants.puuid}, #{matchId}, 'stat_flex', 2, #{participants.perks.statPerks.flex}),
+      (#{participants.puuid}, #{matchId}, 'stat_defense', 3,
+      #{participants.perks.statPerks.defense})
```
잘은 모르겠지만 1,2,3 같은 하드코딩 된 값들도 앞에서 파라미터로 받아야 나중에 변경될때 대응이 수월하지 않을까요?
자바코드에서 넘길때 주석으로 설명도 가능할것 같구요~

f-lab-kyle commented on src/test/java/com/nooblol/account/service/impl/MatchGameInfoServiceImplTest.java:
```diff
@@ -0,0 +1,178 @@
+  String responseOkMatchId = "KR_6077331700";
+
+  @Test
+  @DisplayName("matchId 전달시 RIOT API의 {matchId}가 정상 치환이 되고 URI주소가 정상적으로 Return되는지 확인")
```
```suggestion
  @DisplayName("matchId 전달시 RIOT API의 {matchId}가 정상 치환이 되고 URI주소를 획득 가능하다.")
```

f-lab-kyle commented on src/test/java/com/nooblol/account/service/impl/MatchGameInfoServiceImplTest.java:
```diff
@@ -0,0 +1,178 @@
+    try {
+      assertEquals(matchGameInfoService.insertMatchDataByDB(dto), false);
+    } catch (Exception e) {
+      e.printStackTrace();
```
이런 녀석들은 로깅을 로거(e.g. slf4j )에 의해 남기는게 좋을것 같아요

f-lab-kyle reviewed (commented):
테스트에 힘쓰신게 보이네요 ㅎㅎ
mocking할때 도움주는 https://github.com/naver/fixture-monkey 이런녀석들도 있으니 참고 해보세요~

donsonioc2010 committed:
refactor : Riot에서 최근 MatchId List를 수신하는 경우 HttpStatus값이 OK가 아니면 먼저 Return되도록 수정

donsonioc2010 committed:
refactor : NPE를 예방하기 위하여 Optional로 수정

- MatchGameInfoServiceImpl - getMatchDataByRiot(String) 메소드에서 Null을 반환받는 경우 NPE를 예방하기 위한 수정

donsonioc2010 committed:
refactor : existsMatchIdList의 쿼리작업 DB -> Java에서 작업하는 것으로 수, List확인 ObjectUtils.isEmpty()로 변경

donsonioc2010 committed:
test : Display의 문구수정, 로그 추가

donsonioc2010 committed:
refactor : Runes을 입력하는 파라미터를 Java에서 제작하여 Mybatis로 던져주도록 수정

donsonioc2010 committed:
refactor : Transaction에 대한 설명추가 및 이후 문제 발생시 Exception으로 예외처리 할 수 있도록 수정
